### PR TITLE
Add 'Upload Files...' menu entries

### LINKED ISSUE #4088: Add a File menu entry to upload files
When you want to upload a file into your Theia IDE, there is no visible option to do that.

Most people will first upload files to somewhere, and then `wget` or `curl` or `git pull` (!) desired files from the Terminal. This is terrible.

Adding a File menu entry that says `Upload file...` and opens a file picker would be awesome for UX.

### PR DESCRIPTION
**Pull request by akosyakov**
fix #4088

TODO:
- [x] fix handling of global selection by the navigator: it does not clean it up on disposal
- [x] relocate & group download & upload in the main menu
- [x] relocate & group download & upload in the navigator context menu
- [x] reveal a folder in the navigator after upload 
- [x] ~open a CQ for~ formidable - it's MIT and does not have 3rd party dependencies, we can wait till the CQ process for prod dependencies is clear
- [x] sign off a commit properly
- [x] uploading to the same folder twice should work

upload

### CODE REVIEW

File: packages/filesystem/src/browser/download/file-download-command-contribution.ts [hunk 1]
```diff
@@ -57,29 +76,20 @@ export class FileDownloadCommandContribution implements CommandContribution {
         return this.isDownloadEnabled(uris);
     }
 
-    protected getUris(uri: Object | undefined): URI[] {
```
Comment (akosyakov) [inline packages/filesystem/src/browser/download/file-download-command-contribution.ts]: @kittaakos i've removed it, seems to be just left overs which were never used

**Comment by akosyakov** [comment]
@jankeromnes @jgbradley1 Could you try it out?

**Comment by akosyakov** [comment]
@vzhukovskii There was a bug in the navigator and plugin tree views that the global selection was not cleaned properly when they are disposed. I've introduced `TreeWidgetSelection` to deal with that in `TreeWidget`. One can enable it via `globalSelection` props. Could you please have a look that it does not break menu items for tree view? I would it as well.

**Comment by vzhukovskii** [comment]
@akosyakov thanks! Will try to check it asap.

**Comment by vzhukovskii** [comment]
Looks like it doesn't break existed logic on plugin side.

**Comment by jgbradley1** [comment]
Nice. I like it! One minor UI issue. If a user has a directory open (i.e. by clicking on the drop down arrow) in the workspace and uploads files to it, they will not actually see the file structure updated with the new file names. You must first close the folder and reopen it.

Is there a function call to refresh the file tree after the upload process is complete? This same issue exist when you delete a file. You have to close the folder and open it back up again to see the changes occur.

**Comment by jgbradley1** [comment]
This is not a necessary feature right now, but we should think about some way to communicate upload progress on a status bar or on the blue bar at the bottom of the webpage.

**Comment by akosyakov** [comment]
> This is not a necessary feature right now, but we should think about some way to communicate upload progress on a status bar or on the blue bar at the bottom of the page.

It is on our list to allow blueish progress bar for different top-level widgets similar what you see in VS Code or Chrome while loading the page. It is even more important for cloud solution than for VS Code, since connection delay and issues is something which we need to deal with.

**Comment by akosyakov** [comment]
>  If a user has a directory open (i.e. by clicking on the drop down arrow) in the workspace and uploads files to it, they will not actually see the file structure updated with the new file names. You must first close the folder and reopen it.

It is fs watching issue. I will have a look into it. A library which we use has troubles with symlinks and does not report some events for real paths :( Refreshing for each operation is work around, it should happen automatically based on fs events.

**Comment by akosyakov** [comment]
- [ ] found a bug that uploading to the same folder twice does not work

**Comment by akosyakov** [comment]
@jgbradley1 could you try again, i've added a fix for fs watching with symlinks

File: packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts [hunk 2]
```diff
@@ -192,6 +192,28 @@ export class NsfwFileSystemWatcherServer implements FileSystemWatcherServer {
...
+            try {
+                // file does not exist try to resolve directory
+                return paths.join(fs.realpathSync(directory), file);
+            } catch {
+                // directory does not exist fall back to symlink
+                return path;
+            }
+        }
+    }
+    protected realpath(path: string): string {
```
Comment (AlexTugarev) [inline packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts]: left over?

File: packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts [hunk 3]
```diff
@@ -111,17 +111,17 @@ export class NsfwFileSystemWatcherServer implements FileSystemWatcherServer {
...
-                    this.pushAdded(watcherId, paths.join(event.directory, event.file!));
+                    this.pushAdded(watcherId, this.resolvePath(event.directory, event.file!));
                 }
                 if (event.action === nsfw.actions.DELETED) {
-                    this.pushDeleted(watcherId, paths.join(event.directory, event.file!));
+                    this.pushDeleted(watcherId, this.resolvePath(event.directory, event.file!));
                 }
                 if (event.action === nsfw.actions.MODIFIED) {
-                    this.pushUpdated(watcherId, paths.join(event.directory, event.file!));
+                    this.pushUpdated(watcherId, this.resolvePath(event.directory, event.file!));
```
Comment (AlexTugarev) [inline packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts]: Q: shouldn't both be reported as changed? I mean the link and the target.

File: packages/filesystem/package.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
     "@theia/core": "^0.4.0",
     "@types/base64-js": "^1.2.5",
     "@types/body-parser": "^1.17.0",
+    "@types/formidable": "^1.0.31",
```
Comment (AlexTugarev) [inline packages/filesystem/package.json]: do we need a CQ for this?

File: packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts (hunk 3 above)
Comment (akosyakov) [reply packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts]: i would need to check whether it is possible to report for all symlins, if not then we stick only with the real part. reportin fs events should be fast, detecting all symlinks would harm perf

**Comment by akosyakov** [comment]
i think we can wait till someone opens bugs for pseudo files and silent override. What should be good default `maxFileSize`?

File: packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts (hunk 3 above)
Comment (AlexTugarev) [reply packages/filesystem/src/node/nsfw-watcher/nsfw-filesystem-watcher.ts]: I checked with vscode behavior, and this change would make it quite similar.

File: packages/filesystem/package.json (hunk 4 above)
Comment (AlexTugarev) [reply packages/filesystem/package.json]: verified it's MIT. thanks!

**Comment by akosyakov** [comment]
@AlexTugarev I've added proper error handling and `uploadMaxFileSize` env variable, by default it is 2048MB. Could you check again?

@jgbradley1 i've also added upload progress reporting :)

**Comment by elaihau** [comment]
Nice feature. i love it. 
I tried uploading different versions of the same file to the same location, and found the newer version replaces the older - that's what i expected. would it be helpful if theia warns the user the older version would be overwritten ?

**Comment by akosyakov** [comment]
> I tried uploading different versions of the same file to the same location, and found the newer version replaces the older - that's what i expected. would it be helpful if theia warns the user the older version would be overwritten ?

it is not easy to implement, this part of code cannot provide responses to a user, we can only fail the whole request.

**Comment by jgbradley1** [comment]
@akosyakov This is great. +1 on the progress bar. Something is wrong with the max file size. It's off by an order of magnitude. I cannot upload any file > 2 MB and they all fail with the `Payload too large` message (tested on Gitpod). Files < 2 MB upload successfully.  The math in your code looks correct however.

`2048 * 1024 * 1024 (bytes) = 2 GB`

**Comment by akosyakov** [comment]
@jgbradley1 good catch! thank you for checking

**Comment by akosyakov** [comment]
@AlexTugarev @jgbradley1 I've fixed and retested locally handling of exceeded payload. It should be fine now. Good to merge?

**Comment by akosyakov** [comment]
merging, @AlexTugarev said that he is fine with changes in off-line conversation

**Pull request merged**